invenio_records_resources/services/records/queryparser/query.py

```diff
@@ -61,19 +61,55 @@ class SearchOptions:
             )
     """
 
-    def __init__(self, identity=None, extra_params=None, tree_transformer_factory=None):
+    def __init__(self, identity=None, extra_params=None, tree_transformer_cls=None):
         """Initialise the parser."""
         self.identity = identity
+        self.tree_transformer_cls = tree_transformer_cls
         self.extra_params = extra_params or {}
-        self.tree_transformer_factory = tree_transformer_factory
+        # the pop or {} is needed due to extra_params being passed from the factory
+        # it is possible that e.g. allow_list=None and then it will fail to set()
+        self.mapping = self.extra_params.pop("mapping", None) or {}
+        self.allow_list = self._calculate_allow_list(
+            self.extra_params.pop("allow_list", None), self.mapping
+        )
+        # process fields list
+        self.fields = self._calculate_fields(
```

Comment:
small suggestion: does it make sense if `fields` is a cached property? In the current status, it's only used when the query defaults to `multi_match`.

Reply:
It can be made a `property`, that is true. However, I'm slightly against cached property, we've had some interesting side effects in the past (that's why I made it a field).